torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -915,6 +917,275 @@ Stmt* FlattenIndexes(Stmt* s) {
+  Store* st = getStoreStmtOfProducer(s);
+  if (!st) {
+    return;
+  }
+
+  // Infer bounds info for all accesses that we make in the loop
+  auto loop_bounds_info = inferBounds(f->body());
+
+  // store_bounds_info holds bounds info for the store we're trying to move to
+  // the loop. If its result isn't accessed in the loop at all - do nothing and
+  // exit early.
+  TensorAccessBoundsInfo store_bounds_info;
+  bool found = false;
+  for (const TensorAccessBoundsInfo& p : loop_bounds_info) {
+    if (p.buf == st->buf()) {
+      store_bounds_info = p;
+      found = true;
+    }
+  }
+  if (!found) {
+    return;
+  }
+
+  // Compute dimensions of the temp buffer we would need to allocate
+  std::vector<const Expr*> dims;
+  for (size_t i = 0; i < store_bounds_info.start.size(); i++) {
+    const Expr* dim = IRSimplifier::simplify(new Add(
+        new Sub(store_bounds_info.stop[i], store_bounds_info.start[i]),
+        new IntImm(1)));
+    dims.push_back(dim);
+  }
+
+  // TODO: Use name-hint of the producer instead of "temp"
+  const Buf* temp_buf =
+      new Buf(new Var("temp", store_bounds_info.buf->dtype()), dims);
+
+  // Generate index variables for 'temp'
+  std::vector<const Expr*> temp_indices(dims.size());
+  for (size_t i = 0; i < dims.size(); i++) {
+    // TODO: Use name-hint of the producer indices instead of 'idx'
+    temp_indices[i] = new Var(std::string("idx") + std::to_string(i), kInt);
+  }
+
+  // Prepare substitute rules for constructing the temp statement from the prod
+  // statement
+  // TODO: Instead of going up the loop nest we should go through the indices in
+  // the original tensor expression. The loops in the nest might've been
+  // modified (e.g. split or merged) so that the loop indices no longer
+  // correspond to the indices of the original expression and even their number
+  // might be different. In that case, the loop below would crash.
+  std::vector<const Var*> prod_indices = getOuterLoopIndexes(s);
+  std::vector<std::pair<const Var*, const Expr*>> rewrite_indices_map;
+  for (size_t i = 0; i < prod_indices.size(); i++) {
+    const Expr* offset = store_bounds_info.start[i];
+    rewrite_indices_map.push_back(
+        {prod_indices[i], new Add(temp_indices[i], offset)});
+  }
+  // Construct the temp statement
+  Stmt* bd = new Store(
+      temp_buf,
+      temp_indices,
+      Substitute(st->value(), rewrite_indices_map),
+      st->mask());
+
+  // Construct the loop nest for the temp computation
+  for (size_t i = 0; i < dims.size(); i++) {
+    // We're creating loops from innermost to outermost, so we need to access
+    // dimensions in reversed order.
+    size_t dim_idx = dims.size() - 1 - i;
+    bd = new For(
+        dynamic_cast<const Var*>(temp_indices[dim_idx]),
+        new IntImm(0),
+        dims[dim_idx],
+        bd);
+  }
+
+  // Generate alloc/free stmts for 'temp'
+  Stmt* al = new Allocate(temp_buf->base_handle(), st->value()->dtype(), dims);
```

Comment:
I am still thinking whether the temp should be here, or near the global scope. Maybe a TODO for us to revisit this later?